business/transaction_service/charge.go

```diff
@@ -2,11 +2,135 @@ package transaction_service
 
 import (
 	"context"
+	"errors"
+	"fmt"
+	"time"
 
 	"mock-payment-provider/business"
+	"mock-payment-provider/primitive"
+	"mock-payment-provider/repository"
 )
 
 func (d Dependency) Charge(ctx context.Context, request business.ChargeRequest) (business.ChargeResponse, error) {
-	// TODO implement me
-	panic("implement me")
+	// Validate the transaction amount and the amount of each product items
+	var totalAmount int64
+	for _, item := range request.ProductItems {
+		totalAmount += item.Price * item.Quantity
+	}
+
+	if totalAmount != request.TransactionAmount {
+		return business.ChargeResponse{}, business.ErrMismatchedTransactionAmount
+	}
+
+	switch request.PaymentType {
+	case primitive.PaymentTypeVirtualAccountBCA:
+		fallthrough
+	case primitive.PaymentTypeVirtualAccountMandiri:
+		fallthrough
+	case primitive.PaymentTypeVirtualAccountBRI:
+		fallthrough
+	case primitive.PaymentTypeVirtualAccountBNI:
+		// Create new transaction
+		err := d.TransactionRepository.Create(
+			ctx,
+			request.OrderId,
+			request.TransactionAmount,
+			request.PaymentType,
+			primitive.TransactionStatusPending,
+			time.Now().Add(time.Hour*24),
+		)
```

Comment:
Why not pass into `struct`, instead of by `parameters`?
Having a lot of parameters might make the code look less good?